### PR TITLE
Document fast data path feature

### PR DESCRIPTION
**Pull request by awh**
This is a replacement for #1510 that is targeted against master.

### CODE REVIEW

**Comment by awh** [comment]
@rade @dpw I've added some user facing documentation, would appreciate review.

File: site/features.md [hunk 1]
```diff
@@ -55,6 +56,28 @@ now re-use the same tools and techniques when deploying applications
...
 our data centre.
 
+### <a name="automatic-overlay-selection"></a>Automatic overlay selection
+
+Weave supports two overlay technologies - a custom UDP encapsulation
+('sleeve') featuring optional encryption and underlay PMTU discovery,
+and a VXLAN transport ('fastdp') that leverages the Open vSwitch
+kernel module to provide near-native performance. When connecting to a
+new peer, the router automatically prefers the fastdp overlay under
+the following conditions:
```
Comment (rade) [inline site/features.md]: That's a lot of technical jargon. Overlay? UDP encapsulation? PMTU discovery? VXLAN? Open vSwitch kernel module? ... I don't think a user should have to know about any of these.

How about we call this section "Optimised data path". Explain in a sentence what we mean by "data path". Explain that normally weave will use a "fast in-kernel data path" or s.t. like that.

File: site/features.md [hunk 2]
```diff
@@ -55,6 +56,28 @@ now re-use the same tools and techniques when deploying applications
...
+
+* UDP port 6784 must be reachable (in addition to TCP/UDP 6783)
```
Comment (rade) [inline site/features.md]: We need to mention the 6784 in the README (where we mention 6783). Needs a separate PR on a readme-v1.2 (or whatever naming scheme we used previously).

File: site/features.md [hunk 3]
```diff
@@ -282,7 +305,8 @@ way to generate a random password which satsifies this requirement is
 
     < /dev/urandom tr -dc A-Za-z0-9 | head -c9 ; echo
 
-The same password must be specified for all weave peers.
+The same password must be specified for all weave peers. Note that
+supplying a password will disable support for the fastdp overlay.
```
Comment (rade) [inline site/features.md]: link back to the new section.

File: site/features.md [hunk 4]
```diff
@@ -453,7 +477,9 @@ To enable this, the network must be configured to permit TCP and UDP
 connections to the weave port of the docker hosts. The weave port
 defaults to 6783. This can be overriden by setting `WEAVE_PORT`, but
 it is highly recommended that all peers in a weave network are given
-the same port setting.
+the same port setting. If you wish to use the fastdp overlay, the UDP
+port with value `WEAVE_PORT+1` (6784 by default) must also be
+permitted.
```
Comment (rade) [inline site/features.md]: link back to the new section.

File: site/features.md [hunk 5]
```diff
@@ -55,6 +56,28 @@ now re-use the same tools and techniques when deploying applications
...
+Weave supports two overlay technologies - a custom UDP encapsulation
+('sleeve') featuring optional encryption and underlay PMTU discovery,
+and a VXLAN transport ('fastdp') that leverages the Open vSwitch
+kernel module to provide near-native performance. When connecting to a
+new peer, the router automatically prefers the fastdp overlay under
+the following conditions:
+
+* UDP port 6784 must be reachable (in addition to TCP/UDP 6783)
+* Underlay network PMTU must be greater than or equal to 1460 bytes
+* Encryption must be disabled
```
Comment (rade) [inline site/features.md]: "not be enabled" (given that it is an opt-in)

File: site/features.md (hunk 1 above)
Comment (dpw) [inline site/features.md]: I don't think we should use the term "overlay" at all.

"VXLAN transport" risks creating the expectation of VXLAN interoperability. At most "... that leverages the Open vSwitch kernel module and VXLAN encapsulation ...".

**Comment by awh** [comment]
I have removed the information content as requested.

File: site/features.md (hunk 2 above)
Comment (awh) [inline site/features.md]: OK I will split that out of here once we're happy overall.

File: site/features.md [hunk 6]
```diff
@@ -55,6 +56,20 @@ now re-use the same tools and techniques when deploying applications
 as containers as we would have done when deploying them 'on metal' in
 our data centre.
 
+### <a name="optimised-traffic-transport"></a>Optimised traffic transport
```
Comment (rade) [inline site/features.md]: s/traffic/data?

Comment (dpw) [inline site/features.md]: What's wrong with calling this section 'fast datapath'?  If that name is not suitable for a section title, we should find a name we like better and stick to it (fixing 'fastdp' in the status and WEAVE_NO_FASTDP accordingly).

File: site/features.md [hunk 7]
```diff
@@ -55,6 +56,20 @@ now re-use the same tools and techniques when deploying applications
...
+
+Weave automatically chooses the fastest available method to transport
+traffic between peers. The most performant of these ('fastdp') offers
+near-native throughput and latency but does not support encryption;
+consequently supplying a password will cause the router to fall back
+to a slower mode ('sleeve') that does. You can see which method is in
```
Comment (rade) [inline site/features.md]: "...does not support encryption. The slower method ('sleeve') is selected when [encryption is enabled](#security), or communication via the 'fastdp' method fails for some other reason. You can see..."

File: site/features.md (hunk 6 above)
Comment (rade) [inline site/features.md]: I would be fine with "Fast data path". It's perhaps not the most descriptive we can come up with, but it's not far off.

**Comment by awh** [comment]
* 'Optimised traffic transport' becomes 'Optimised data transport'
* Added explanation of sleeve fallback under adverse network conditions
* README changes pulled out in #1541

File: site/troubleshooting.md [hunk 8]
```diff
@@ -161,8 +161,9 @@ The columns are as follows:
     * `pending` - TCP connection up, waiting for confirmation of
        UDP heartbeat
     * `established` - TCP connection and corresponding UDP path are up
- * Info - the remote peer name and nickname for (un)established
-   connections, the failure reason for failed and retrying connection
+ * Info - the traffic transport method, remote peer name and nickname
+   for (un)established connections, the failure reason for failed and
+   retrying connection
```
Comment (rade) [inline site/troubleshooting.md]: I know this was wrong before, but there is no such thing as "unestablished". Should say "pending and established". Also, I reckon "failed and retrying" should come before "pending and established", to match the order in Status.

**Comment by awh** [comment]
* 'Optimised data transport' becomes 'Fast data path'

File: site/troubleshooting.md (hunk 8 above)
Comment (awh) [inline site/troubleshooting.md]: Done.

**Comment by awh** [comment]
@dpw did you have any comments on the high level overview doc?

**Comment by dpw** [comment]
> @dpw did you have any comments on the high level overview doc?

I have a fairly strong idea of what a "weave router internals" doc should look like, as I was planning to write one as noted in my "things still to do" list on the fastdp PR.  But I appreciate that you had to figure stuff out as you went along, and I don't see any harm in having that written down.  Your doc might be quite far from what I had in mind though, so I'm not sure I will have much useful feedback.

If you want to collaborate to put together a comprehensive doc, we can work on that.  Or not.  Either is fine by me.  In the former case, I'd suggest splitting the user-facing documentation part out onto a separate PR so we can get that merged promptly.

**Comment by awh** [comment]
I've pulled the overview out of this PR; only updates to the user facing documentation remain.

**Pull request merged**